### DISCUSSION TITLE
concurrent mofification on page

### DISCUSSION THREAD

**Discussion by kran** [ArcadeData/arcadedb#475] [Q&A] [answered]
`Concurrent modification on page PageId(25/8) in file 'stats_user_0.25.65536.v0.bucket' (current v.102 <> database v.103). Please retry the operation (threadId=167)`

I use arcadedb in embedded mode, and got errors above in multiple threads code
how to solve this ?

**Comment by lvca** [comment, marked as answer]
ArcadeDB (like OrientDB and other DBMSs) uses a no-lock optimistic approach to transactions. If the same page is modified by another thread (or client), then the transaction must be repeated. You can wrap your transaction with:

```java
database.transaction( ()-> {
  ...
});
```

By default the transaction block is repeated in case of conflict up to 3 times. This setting can be changed by setting the `arcadedb.txRetries` to the default number of retries.

You can also decide the number of retries by using this method:

```java
database.transaction( ()-> {
  ...
}, false, 10);
```

Where 10 is the maximum number of attempts.

**Reply by sharanAmit** [reply]
I also received the same error , while connecting with python using http API. What should I do?

**Reply by lvca** [reply]
if this is the error, you should just retry the whole transaction. This is to avoid locks on the database and improve concurrency
